File: Python/WebDev/WebServer/library/TcpListener.py
import socket, threading

class TcpListener:
    def __init__(self, ipAddr, port):
        self.__ipAddr = ipAddr
        self.__port = port
        self.__server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.__running = False

    # override this
    def clientConnected(self, clientSock, clientAddr):
        pass

    # override this
    def clientDisconnected(self, clientSock, clientAddr):
        pass

    # override this
    def msgReceived(self, clientSock, clientAddr, msg):
        pass

    # override this
    def serverEvent(self, msg):
        pass

    # don't overrride
    def acceptThread(self):
        self.__running = True

        while self.__running:
            self.__server.listen(1)
            clientSock, clientAddr = self.__server.accept()
            thread = threading.Thread(target=self.clientThread, args=(clientSock, clientAddr))
            thread.start()

    # don't override
    def clientThread(self, clientSock, clientAddr):
        self.clientConnected(clientSock, clientAddr)
        msg = ''
        while True:
            data = clientSock.recv(2048)
            msg = data.decode()
            if not msg:
                break
            elif msg.isspace():
                continue
            self.msgReceived(clientSock, clientAddr, msg)
            self.send(clientSock, msg)
        clientSock.close()
        self.clientDisconnected(clientSock, clientAddr)

    # don't override
    def send(self, clientSock, msg):
        clientSock.send(bytes(msg,'UTF-8'))

    def stop(self):
        self.__running = False
        self.__incomingConnectionThread.join()
        self.serverEvent("Server will close when all clients disconnected")
        self.__server.close()

    def open(self):
        self.__server.bind((self.__ipAddr, self.__port))

        self.serverEvent("Server started at " + self.__ipAddr + ":" + str(self.__port))
        
        self.__incomingConnectionThread = threading.Thread(target=self.acceptThread)
        self.__incomingConnectionThread.start()
        

File: Java/Networking/src/com/networking/sockets/TcpListener.java
package com.networking.sockets;

import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import com.networking.util.Logger;

public class TcpListener {
	protected int port;
	protected String ipAddr_str;
	protected InetAddress ipAddr;
	protected Logger logger;
	protected ServerSocket socket;
	protected boolean running;
	protected boolean bound;
	protected List<TcpClient> clients;
	
	private Scanner scanner;
	
	public TcpListener(int port, String ipAddr_str, Logger logger)  {
		this.port = port;
		this.ipAddr_str = ipAddr_str;
		this.logger = logger;
		this.running = false;
		this.bound = false;
	}
	
	public final void bind() throws IOException {
		this.ipAddr = InetAddress.getByName(ipAddr_str);
		this.socket = new ServerSocket(this.port, -1, this.ipAddr);
		clients = new ArrayList<>();
		this.running = false;
		this.bound = true;
		
		logger.Log("Server bound");
	}
	
	public void run() throws IOException {
		if (!running && bound) {
			this.running = true;
			
			logger.Log("Server started");
			
			Runnable cmdRunnable = () -> {
				commandThread();
			};
			Thread cmdThread = new Thread(cmdRunnable);
			cmdThread.start();
			
			while (this.running) {
				Socket client = socket.accept();
				TcpClient tcpClient = new TcpClient(client);
				clients.add(tcpClient);
				Runnable clientRunnable = () -> {
					clientThread(tcpClient);
				};
				Thread clientThread = new Thread(clientRunnable);
				
				clientThread.start();
				onClientConnected(client);
			}
			
			for (TcpClient client : clients) {
				if (client.socket.isConnected()) {
					client.socket.close();
				}
			}
			
			socket.close();
			
			running = false;
			bound = false;
			
			logger.Log("Server stopped");
		}
	}
	
	public final void send(TcpClient receiver, String msg) throws IOException {
		receiver.send(msg);
	}
	
	public final void broadcast(String msg) throws IOException {
		for (TcpClient client : clients) {
			send(client, msg);
		}
	}
	
	public final TcpClient getClient(Socket sock) {
		for (TcpClient client : clients) {
			if (client.socket == sock) {
				return client;
			}
		}
		
		return null;
	}
	
	public void stopServer() throws IOException {
		this.running = false;
		// start new client and connect to get to next iteration of main loop
		Socket socket = new Socket(ipAddr_str, port);
		socket.close();
	}
	
	private void clientThread(TcpClient client) {
		while (this.running && client.socket.isConnected()) {
			try {
				String input = client.reader.readLine();
				if (input == null) {
					clients.remove(client);
					onClientDisconnected(client.socket);
					break;
				}
				client.send("Msg received");
				onMessageReceived(client.socket, input);
			} catch (Exception e) {
				e.printStackTrace();
				clients.remove(client);
				onClientDisconnected(client.socket);
				break;
			}
		}
	}
	
	private void commandThread() {
		scanner = new Scanner(System.in);
		
		while (this.running) {
			String cmd = scanner.nextLine();
			
			try {
				switch (cmd) {
				case "stop":
					this.running = false;
					// start new client and connect to get to next iteration of main loop
					Socket socket = new Socket(ipAddr_str, port);
					socket.close();
					break;
				default:
					broadcast(cmd);
					break;
				}
			} catch (Exception e) {
				e.printStackTrace();
				return;
			}
		}
	}
	
	protected void onClientConnected(Socket client) {
		logger.Log("client connected");
	}
	
	protected void onClientDisconnected(Socket client) {
		logger.Log("client disconnected");
	}
	
	protected void onMessageReceived(Socket sender, String msg) {
		logger.Log("msg received: " + msg);
	}
}